ekncontent/ekncontent/eknc-engine.c

```diff
@@ -0,0 +1,508 @@
+ * @error: #GError for error reporting.
+ *
+ * Finishes a call to eknc_engine_query. Returns a #EkncQueryResults object.
+ * Throws an error if one occurred.
+ *
+ * Returns: (transfer full): the results object
+ */
+EkncQueryResults *
+eknc_engine_query_finish (EkncEngine *self,
+                          GAsyncResult *result,
+                          GError **error)
+{
+  g_return_val_if_fail (EKNC_IS_ENGINE (self), NULL);
+  g_return_val_if_fail (G_IS_TASK (result), NULL);
+  g_return_val_if_fail (error == NULL || *error == NULL, NULL);
+
+  return g_task_propagate_pointer (G_TASK (result), error);
+}
+
+/**
+ * eknc_engine_get_domain:
+ * @self: the engine
+ * @error: #GError for error reporting.
+ *
+ * Get a #EkncDomain object for the default-app-id. Requires the default-app-id
+ * property to be set.
+ *
+ * Returns the domain if it has already been created, and synchronously creates
+ * one if none exists. Returns an error if one occurred while initializing the
+ * domain.
+ *
+ * Returns: (transfer none): the default domain
+ */
+EkncDomain *
+eknc_engine_get_domain (EkncEngine *self,
+                        GError **error)
+{
+  g_return_val_if_fail (EKNC_IS_ENGINE (self), NULL);
+
+  return eknc_engine_get_domain_for_app (self, self->default_app_id, error);
+}
+
+/**
+ * eknc_engine_get_domain_for_app:
+ * @self: the engine
+ * @app_id: the id of the application to load the object from
+ * @error: #GError for error reporting.
+ *
+ * Get a #EkncDomain object for the a given app id.
+ *
+ * Returns the domain if it has already been created, and synchronously creates
+ * one if none exists. Returns an error if one occurred while initializing the
+ * domain.
+ *
+ * Returns: (transfer none): the domain
+ */
+EkncDomain *
+eknc_engine_get_domain_for_app (EkncEngine *self,
+                                const gchar *app_id,
+                                GError **error)
+{
+  g_return_val_if_fail (EKNC_IS_ENGINE (self), NULL);
+  g_return_val_if_fail (app_id && *app_id, NULL);
+  g_return_val_if_fail (error == NULL || *error == NULL, NULL);
+
+  if (g_hash_table_contains (self->domains, app_id))
+    return g_hash_table_lookup (self->domains, app_id);
+
+  EkncDomain *domain;
+  if (!(domain = eknc_domain_get_impl (app_id, self->xapian_bridge, NULL, error)))
+    return NULL;
+  g_hash_table_insert (self->domains, (gpointer)app_id, domain);
+  return domain;
+}
+
+/**
+ * eknc_engine_get_default:
+ *
+ * Get the default engine object. Generally you should use this instead of
+ * creating an engine object manually.
+ *
+ * Returns: (transfer none): the default engine
+ */
+EkncEngine *
+eknc_engine_get_default (void)
+{
+  static EkncEngine *engine;
+
+  if (G_UNLIKELY (g_once_init_enter (&engine)))
+    {
+      const gchar *language = eknc_get_current_language ();
+      static EkncEngine *e;
+      e = g_object_new (EKNC_TYPE_ENGINE, "language", language, NULL);
+      g_once_init_leave (&engine, e);
```

Comment:
nitpick: `g_once_init_leave (&engine, g_object_new (EKNC_TYPE_ENGINE, ...))` reads cleaner IMO, but it's totally up to your preference